presto-spark/src/main/java/com/facebook/presto/spark/planner/SparkRddPlanner.java

```diff
@@ -0,0 +1,254 @@
+                return sparkContext.parallelize(serializedRequests)
+                        .flatMapToPair(createTaskProcessor(compilerFactory, taskStatsCollector));
+            }
+
+            List<SubPlanWithTaskSources> children = subPlan.getChildren();
+            checkArgument(
+                    remoteSources.size() == children.size(),
+                    "number of remote sources doesn't match the number of child stages: %s != %s",
+                    remoteSources.size(),
+                    children.size());
+
+            if (children.size() == 1) {
+                // Single remote source
+                SubPlanWithTaskSources childSubPlan = getOnlyElement(children);
+                JavaPairRDD<Integer, byte[]> childRdd = createRdd(childSubPlan);
+
+                PlanFragment childFragment = childSubPlan.getFragment();
+                RemoteSourceNode remoteSource = getOnlyElement(remoteSources);
+                List<PlanFragmentId> sourceFragmentIds = remoteSource.getSourceFragmentIds();
+                checkArgument(sourceFragmentIds.size() == 1, "expected to have exactly only a single source fragment");
+                checkArgument(childFragment.getId().equals(getOnlyElement(sourceFragmentIds)));
+
+                SparkTaskDescriptor sparkTaskDescriptor = createTaskDescriptor(fragment, ImmutableList.of());
+                byte[] serializedTaskDescriptor = sparkTaskRequestJsonCodec.toJsonBytes(sparkTaskDescriptor);
+
+                PartitioningHandle partitioning = fragment.getPartitioning();
+                if (partitioning.equals(COORDINATOR_DISTRIBUTION)) {
+                    // TODO: We assume COORDINATOR_DISTRIBUTION always means OutputNode
+                    // But it could also be TableFinishNode, in that case we should do collect and table commit on coordinator.
+
+                    // TODO: Do we want to return an RDD for root stage? -- or we should consider the result will not be large?
+                    List<Tuple2<Integer, byte[]>> collect = childRdd.collect();
+                    List<Tuple2<Integer, byte[]>> result = ImmutableList.copyOf(compilerFactory.create().compile(
+                            0,
+                            serializedTaskDescriptor,
+                            ImmutableMap.of(remoteSource.getId().toString(), collect.iterator()),
+                            taskStatsCollector));
+                    return sparkContext.parallelizePairs(result);
+                }
+                else if (partitioning.equals(FIXED_HASH_DISTRIBUTION) ||
+                        // when single distribution - there will be a single partition 0
+                        partitioning.equals(SINGLE_DISTRIBUTION)) {
+                    String planNodeId = remoteSource.getId().toString();
+                    return childRdd
+                            // TODO: What's the difference of using
+                            //  partitionBy/mapPartitionsToPair vs. groupBy vs. mapToPair ???
```

Comment:
We can remove the comment now. 

`groupBy` goes into `groupByKey` and will "hold all the key-value pairs for any key in memory":

https://github.com/apache/spark/blob/8d3eed33eee0d6e5d2e4d4907869ca6226d9e37f/core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala#L491-L504

Use `partitionedBy` is the right choice :)